pages/learn/develop/container-contracts.md

```diff
@@ -59,7 +59,7 @@ requires:
 
 __Note__ Any type other than `sw.os` or `sw.supervisor` will result in the contract being invalid.
 
-When deploying a release, if the contract requirements are not met, the release will fail, and all devices in the application will remain on their current release. This default behavior can be overridden by including an [optional label](#optional-contracts) in the application docker-compose.yml file. This label specifies which services with unmet requirements will be ignored, with the release being deployed to all other devices in the application. If there are any existing running containers for services with unmet requirements, they will not be destroyed and continue running the previous release.
+When deploying a release, if the contract requirements are not met, the release will fail, and all devices in the application will remain on their current release. This default behavior can be overridden by including an [optional label](#optional-containers) in the application docker-compose.yml file. This label specifies which services with unmet requirements will be ignored, with the release being deployed to all other devices in the application. If there are any existing running containers for services with unmet requirements, they will not be destroyed and continue running the previous release.
```

Comment:
Its unclear where the contract resolution is done, is this on the builder or on the device. If on the builder, does the builder query all the OS versions. I assume it must be the device that decides these things? since if the device has a target release with a failed optional container, when it does a hostOS upgrade to the requisite version. I assume it would then start running the option container @CameronDiver ?

Reply:
Yeah, it all happens on-device, and as the hup occurs the new valid containers will run.

Reply:
Making sure I understand this (will give this a test later). If I have a release where the contract fails, say the OS is below a certain threshold and the release is not deployed to any devices (so not using optional containers). When I then do the OS upgrade on the device so the contract passes this release is then immediately deployed to all devices? Or does this only work in the optional container case as Shaun details above?

Also, what is hup? host upgrade?

Reply:
I confirmed this behaviour by updating the hostOS locally from a failing to passing version and indeed the optional container named `contract` started up upon reboot.

```
07.11.19 21:12:44 (+0000) Not running containers because of contract violations: contract
07.11.19 21:47:06 (+0000) Supervisor starting
07.11.19 21:47:08 (+0000) Killing service 'frontend sha256:d7c9e8e01da39ce3c3442a727f3aea3531e6562b687ca084a9088060c412f5bf'
07.11.19 21:47:08 (+0000) Killing service 'proxy sha256:44974585ebe63e4010fbbfbe161357c857b2bb4c965a6107c6da243af3a50cb2'
07.11.19 21:47:08 (+0000) Killing service 'data sha256:9c9688ec1e49f5c0c4c939c1bd329c6be01cc2b48b14f3f428161445c68e6253'
.......
07.11.19 21:47:59 (+0000) Downloading delta for image 'registry2.balena-staging.com/v2/66b98834140f6bfbb36920c432457142@sha256:0525c2f751b8d41f665e4eeacd97460f6e2658eb44353c71c39b5d37c2555565'
07.11.19 21:53:49 (+0000) Downloaded image 'registry2.balena-staging.com/v2/66b98834140f6bfbb36920c432457142@sha256:0525c2f751b8d41f665e4eeacd97460f6e2658eb44353c71c39b5d37c2555565'
07.11.19 21:53:49 (+0000) Installing service 'contract sha256:835937c4045a877fad29310369262e97dcfacbec06447fa4219d0ba817c68499'
07.11.19 21:53:52 (+0000) Installed service 'contract sha256:835937c4045a877fad29310369262e97dcfacbec06447fa4219d0ba817c68499'
07.11.19 21:53:52 (+0000) Starting service 'contract sha256:835937c4045a877fad29310369262e97dcfacbec06447fa4219d0ba817c68499'
07.11.19 21:53:56 (+0000) Started service 'contract sha256:835937c4045a877fad29310369262e97dcfacbec06447fa4219d0ba817c68499'
```

Reply:
"When deploying a release, if the contract requirements are not met, the release will fail, and all devices in the application will remain on their current release."

Based on @CameronDiver 's comment about it being verified on-device, I guess it's actually "all devices that don't fulfil the contract will remain on their current release", rather than "all devices in the application" ? :man_shrugging: 

Also, it's not clear what "This label specifies which services with unmet requirements will be ignored, with the release being deployed to all other devices in the application." means.